varname/ignore.py

```diff
@@ -75,181 +68,157 @@ def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
-    @lru_cache()
-    def _source_from_frame(frame: FrameType) -> Source:
-        """Get source object from the given frame"""
-        return Source.for_frame(frame)
-
-    def _check_qualname_by_source(self, source: Source) -> None:
-        """Check if qualname references to multiple objects"""
-        if not source.tree:
-            # there is no way to check
-            return
-        nobj = list(source._qualnames.values()).count(self.ignore[1])
-        # don't check for non-existence, as it could contain wildcards
-        if nobj > 1:
-            raise QualnameNonUniqueError(
-                f"Qualname {self.ignore[1]!r} in "
-                f"{self.ignore[0].__name__!r} refers to multiple objects."
+        modfile = getattr(self.module, '__file__', None)
+        if modfile is not None:
+            check_qualname_by_source(
+                Source.for_filename(modfile, self.module.__dict__),
+                self.module.__name__,
+                self.qualname
             )
 
-    @lru_cache()
-    def _check_qualname_by_frame(self, frame: FrameType) -> None:
-        """Check if qualname references to multiple objects by given frame"""
-        source = IgnoreModuleQualname._source_from_frame(frame)
-        self._check_qualname_by_source(source)
-
     def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
         frame = frameinfos[frame_no].frame
         module = cached_getmodule(frame.f_code)
+
         # Return earlier to avoid qualname uniqueness check
-        if module and module != self.ignore[0]:
+        if module and module != self.module:
             return False
 
-        if not module and (
-                getattr(self.ignore[0], MODULE_IGNORE_ID_NAME, object()) !=
-                frame.f_globals.get(MODULE_IGNORE_ID_NAME, object())
+        if (
+                not module and
+                not frame_matches_module_by_ignore_id(frame, self.module)
         ):
             return False
 
-        if not self.qname_checked: # // todo: coverage
-            self._check_qualname_by_frame(frame)
-            self.qname_checked = True
+        source = Source.for_frame(frame)
+        check_qualname_by_source(source, self.module.__name__, self.qualname)
 
-        source = IgnoreModuleQualname._source_from_frame(frame)
         return fnmatch(
             source.code_qualname(frame.f_code),
-            self.ignore[1]
+            self.qualname
         )
 
     def __repr__(self) -> str:
         return ('<IgnoreModuleQualname'
-                f'({self.ignore[0].__name__!r}, {self.ignore[1]})>')
+                f'({self.module.__name__!r}, {self.qualname})>')
 
-class IgnoreFilenameQualname(IgnoreElem):
+class IgnoreFilenameQualname(IgnoreElem, attrs=['filename', 'qualname']):
     """Ignore calls with given qualname in the module with the filename"""
 
     # check qualname uniqueness in this as well?
+    # But where to find module.__dict__ for Source.for_filename?
```

Comment:
Yes, check it in `match` after checking that filenames match and you have Source.for_frame. The check is cached now so it's not a problem. Don't convert the filename to a module because check_qualname_by_source doesn't strictly need a module name, just some label/description.